### PR TITLE
Generic syncing states: Push

### LINKED ISSUE #1309: Use generic composition instead of inheritance for sync states
Currently, we use inheritance to implement all the entity-specific sync states from their abstract generic base classes.

I think the code and especially the tests could be much simpler if instead we have only a single generic sync state and inject the specific behaviour using interfaces and small objects for the different entities.

Consider this partly a suggestion, I didn't actually make the changes to see if works better. It's more of a hunch really. Other opinions welcome.

### PR DESCRIPTION
**Pull request by simonrozsival**
Closes #1309 

This PR changes the way we implement the push-sync states. There are now no entity-specific states, everything is generic 🎉 

Some changes to the data sources were necessary to make everything work well.

There are quite a few changed files - mainly because a lot of files was 🔥.

I had to add a common interface to the API models - to add an extension method to check if the entity changed or not. We can discuss this change if the reviewer disagrees with it.

Please test that pushing works the way it should (focus mainly on the time entries, user data, and preferences).

I know you will be mad at me when you review this - it is A LOT of changes. I don't think it is possible to split it into several PRs though :/ Feel free to ask me anything and if it should be helpful, I can have a call with you explaining what and why was changed.

Squash whenever

### CODE REVIEW

File: Toggl.Foundation/Sync/States/Push/DeleteLocalEntityState.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+            this.dataSource = dataSource;
+        }
+
+        public IObservable<ITransition> Start(TThreadsafeModel entity)
+            => delete(entity)
+                .Select(_ => Deleted.Transition())
+                .Catch((Exception e) => Observable.Return(DeletingFailed.Transition()));
+
+        private IObservable<Unit> delete(TThreadsafeModel entity)
+            => entity == null
```
Comment (kornelijepetak) [inline Toggl.Foundation/Sync/States/Push/DeleteLocalEntityState.cs]: Interestingly enough, the compiler is unable to infer here that `TThreadSafeModel` must be a `class` so it warns about possible comparison of value types with `null`. Strange stuff. Constraints are hard 😛

File: Toggl.Foundation/Sync/States/Push/BasePushEntityState.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+        protected Func<Exception, IObservable<ITransition>> Fail(TThreadsafeModel entity)
+            => exception => shouldRethrow(exception)
+                ? Observable.Throw<ITransition>(exception)
+                : Observable.Return(failTransition(entity, exception));
+
+        private bool shouldRethrow(Exception e)
+            => e is ApiDeprecatedException || e is ClientDeprecatedException || e is UnauthorizedException || e is OfflineException;
+
+        private ITransition failTransition(TThreadsafeModel entity, Exception e)
```
Comment (kornelijepetak) [inline Toggl.Foundation/Sync/States/Push/BasePushEntityState.cs]: Maybe it's just me, but I really dislike nested LINQ queries and nested ternary conditionals.
I would find this more readable (albeit less terse):

```
private ITransition failTransition2(TThreadsafeModel entity, Exception e)
{
    var result = (e, entity);

    if (e is ServerErrorException)
        return ServerError.Transition(result);

    if (e is ClientErrorException)
        return ClientError.Transition(result);

    return UnknownError.Transition(result);
}
```

File: Toggl.Foundation/Sync/States/Push/DeleteLocalEntityState.cs (hunk 1 above)
Comment (kornelijepetak) [inline Toggl.Foundation/Sync/States/Push/DeleteLocalEntityState.cs]: Interestingly enough, the compiler is unable to infer here that `TThreadSafeModel` must be a `class` so it warns about possible comparison of value types with `null`. Strange stuff. Constraints are hard 😛

File: Toggl.Multivac/Extensions/ApiModelExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+               && originalApiModel is ILastChangedDatable originalLastChangeDatable
+                ? originalLastChangeDatable.At > lastChangeDatable.At
+                : !propertiesMatch(apiModel, originalApiModel);
+
+        private static bool propertiesMatch(IApiModel first, IApiModel second)
+        {
+            if (first.GetType() != second.GetType())
+                throw new ArgumentException($"Cannot check if the properties are the same for objects of different types ({first.GetType()}, {second.GetType()})");
+
+            foreach (var property in first.GetType().GetProperties())
```
Comment (kornelijepetak) [inline Toggl.Multivac/Extensions/ApiModelExtensions.cs]: You can replace the `foreach` and the `return` by:

```
return first.GetType()
    .GetProperties()
    .All(p => p.GetValue(first, null).Equals(p.GetValue(second, null)));
```

File: Toggl.Foundation/Sync/States/Push/BasePushEntityState.cs (hunk 2 above)
Comment (simonrozsival) [reply Toggl.Foundation/Sync/States/Push/BasePushEntityState.cs]: Please make an issue and we can have a discussion there if you think this should be changed - I don't want to change the internal implementation of the states in this PR.

File: Toggl.Foundation/Sync/States/Push/DeleteLocalEntityState.cs (hunk 1 above)
Comment (simonrozsival) [reply Toggl.Foundation/Sync/States/Push/DeleteLocalEntityState.cs]: The compiler is probably right - `TDatabaseModel` can be an interface, `TDatabaseModel` can be a struct implementing that interface. I will add the constraint here.

File: Toggl.Foundation/TogglSyncManagerFactory.cs [hunk 4]
```diff
@@ -156,7 +156,7 @@ public static class TogglSyncManager
 
         private static IStateResult configurePushTransitionsForTags(
```
Comment (amulware) [inline Toggl.Foundation/TogglSyncManagerFactory.cs]: There is some duplication with these methods now. Maybe that can be made generic and shared?

File: Toggl.Multivac/Models/IApiModel.cs [hunk 5]
```diff
@@ -0,0 +1,6 @@
+﻿namespace Toggl.Multivac.Models
+{
+    public interface IApiModel
+    {   
+    }
+}
```
Comment (amulware) [inline Toggl.Multivac/Models/IApiModel.cs]: 👀

File: Toggl.Multivac/Extensions/ApiModelExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using Toggl.Multivac.Models;
+
+namespace Toggl.Multivac.Extensions
+{
+    public static class ApiModelExtensions
+    {
+        public static bool HasChanged(this IApiModel apiModel, IApiModel originalApiModel)
```
Comment (amulware) [inline Toggl.Multivac/Extensions/ApiModelExtensions.cs]: This I'm sort of ok with, though it's not ideal.

File: Toggl.Multivac/Extensions/ApiModelExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+                : !propertiesMatch(apiModel, originalApiModel);
+
+        private static bool propertiesMatch(IApiModel first, IApiModel second)
+        {
+            if (first.GetType() != second.GetType())
+                throw new ArgumentException($"Cannot check if the properties are the same for objects of different types ({first.GetType()}, {second.GetType()})");
+
+            foreach (var property in first.GetType().GetProperties())
+            {
+                if (!property.GetValue(first, null).Equals(property.GetValue(second, null)))
```
Comment (amulware) [inline Toggl.Multivac/Extensions/ApiModelExtensions.cs]: But this...

File: Toggl.Multivac/Extensions/ApiModelExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+using Toggl.Multivac.Models;
+
+namespace Toggl.Multivac.Extensions
+{
+    public static class ApiModelExtensions
+    {
+        public static bool HasChanged(this IApiModel apiModel, IApiModel originalApiModel)
+            => apiModel is ILastChangedDatable lastChangeDatable
+               && originalApiModel is ILastChangedDatable originalLastChangeDatable
+                ? originalLastChangeDatable.At > lastChangeDatable.At
```
Comment (amulware) [inline Toggl.Multivac/Extensions/ApiModelExtensions.cs]: Is this comparison not inverted?

File: Toggl.Multivac/Extensions/ApiModelExtensions.cs (hunk 6 above)
Comment (amulware) [reply Toggl.Multivac/Extensions/ApiModelExtensions.cs]: There need to be tests though.

File: Toggl.Multivac/Extensions/ApiModelExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+                : !propertiesMatch(apiModel, originalApiModel);
```
Comment (amulware) [inline Toggl.Multivac/Extensions/ApiModelExtensions.cs]: If the properties are the same, what do we care if it gets overridden or not

File: Toggl.Foundation/Sync/States/Push/Interfaces/IPushEntityState.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using Toggl.Foundation.Models.Interfaces;
+
+namespace Toggl.Foundation.Sync.States.Push.Interfaces
+{
+    public interface IPushEntityState<TThreadsafeModel>
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Foundation/Sync/States/Push/Interfaces/IPushEntityState.cs]: Exactly.

File: Toggl.Foundation/Sync/States/Push/PushState.cs [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+                        ? (ITransition)PushEntity.Transition(entity)
+                        : NothingToPush.Transition());
+
+        private IObservable<TThreadsafeModel> getOldestUnsynced()
+            => dataSource
+                .GetAll(entity => entity.SyncStatus == SyncStatus.SyncNeeded)
+                .SingleAsync()
+                .Select(entities => entities
+                    .OrderBy(entity => entity.At)
+                    .FirstOrDefault());
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Foundation/Sync/States/Push/PushState.cs]: Yeah, sorting the whole list is unnecessary. We should create an issue.

File: Toggl.Foundation/Sync/States/Push/PushSingleState.cs [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+using System.Reactive.Linq;
+using Toggl.Foundation.DataSources.Interfaces;
+using Toggl.Foundation.Models.Interfaces;
+using Toggl.Foundation.Sync.States.Push.Interfaces;
+using Toggl.Multivac;
+using Toggl.PrimeRadiant;
+
+namespace Toggl.Foundation.Sync.States.Push
+{
+    internal sealed class PushSingleState<TDatabaseModel, TThreadsafeModel> : IPushState<TThreadsafeModel>
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Foundation/Sync/States/Push/PushSingleState.cs]: 🤔

File: Toggl.Foundation/Sync/States/Push/UnsyncableEntityState.cs [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+                failedPush.Reason == null
+                    ? nameof(failedPush.Reason)
+                    : nameof(failedPush.Entity)));
+
+        private IObservable<ITransition> failBecauseOfUnexpectedError(Exception reason)
+            => Observable.Throw<Transition<TThreadsafeModel>>(reason);
+
+        private IObservable<ITransition> markAsUnsyncable(TThreadsafeModel entity, string reason)
+            => dataSource
+                .UpdateWithConflictResolution(entity, createUnsyncableFrom(entity, reason), new OverwriteIfLocalDidNotChange<TDatabaseModel, TDatabaseModel>(entity))
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Foundation/Sync/States/Push/UnsyncableEntityState.cs]: 😆

File: Toggl.Foundation/Sync/States/Push/UnsyncableEntityState.cs [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+            Ensure.Argument.IsNotNull(createUnsyncableFrom, nameof(createUnsyncableFrom));
+            
+            this.dataSource = dataSource;
+            this.createUnsyncableFrom = createUnsyncableFrom;
+        }
+
+        public IObservable<ITransition> Start((Exception Reason, TThreadsafeModel Entity) failedPush)
+            => failedPush.Reason == null || failedPush.Entity == null
+                ? failBecauseOfNullArguments(failedPush)
+                : failedPush.Reason is ApiException apiException
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Foundation/Sync/States/Push/UnsyncableEntityState.cs]: I didn't change the logic - it is the original implementation that has been there from the first release. It should be entered only with client exceptions, yes. I am not sure about this implementation detail - maybe we should throw if it is not a client error exception?

[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Foundation/Sync/States/Push/UnsyncableEntityState.cs]: Still it would be good to have a separate issue and PR to change this.

File: Toggl.Foundation/TogglSyncManagerFactory.cs (hunk 4 above)
Comment (simonrozsival) [reply Toggl.Foundation/TogglSyncManagerFactory.cs]: 🤔 .. it might be possible to do, I am not really sure if it is a good idea to do - we have good control of the process for every type of entity. I will look into it and if it is simple, I will change it.

File: Toggl.Multivac/Extensions/ApiModelExtensions.cs (hunk 6 above)
Comment (simonrozsival) [reply Toggl.Multivac/Extensions/ApiModelExtensions.cs]: Do you see any potential problems with this?

I will add tests.

File: Toggl.Multivac/Extensions/ApiModelExtensions.cs (hunk 7 above)
Comment (simonrozsival) [reply Toggl.Multivac/Extensions/ApiModelExtensions.cs]: I already changed this with help from @kornelijepetak to use LINQ and make it much nicer.

File: Toggl.Multivac/Extensions/ApiModelExtensions.cs (hunk 9 above)
Comment (simonrozsival) [reply Toggl.Multivac/Extensions/ApiModelExtensions.cs]: If the properties are the same -> we can simply override it. If they aren't the same -> the entity changed (e.g., preferences) and we don't want to override the local changes - that is why we care. Or did you mean something else by this comment?

File: Toggl.Multivac/Models/IApiModel.cs (hunk 5 above)
Comment (simonrozsival) [reply Toggl.Multivac/Models/IApiModel.cs]: Another empty interface, I know... but I wasn't able to create the extension method without this interface. Also I think it is good if the models in one layer/project have some common ancestor without any exceptions.

File: Toggl.Foundation.Tests/Toggl.Foundation.Tests.csproj [hunk 15]
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Foundation.Tests/Toggl.Foundation.Tests.csproj]: 🙈

File: Toggl.Foundation.Tests/Sync/States/Push/PushStateTests.cs [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+            dataSource.GetAll(Arg.Any<Func<IDatabaseTestModel, bool>>())
+                .Returns(Observable.Throw<IEnumerable<IThreadsafeTestModel>>(new Exception()));
+
+            Action callingStart = () => state.Start().SingleAsync().Wait();
+
+            callingStart.ShouldThrow<Exception>();
+        }
+
+        [Fact, LogIfTooSlow]
+        public void ReturnsPushEntityTransitionWithTheOldestEntity()
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Foundation.Tests/Sync/States/Push/PushStateTests.cs]: I am not sure how you mean it. We always set the at dates.

File: Toggl.Multivac/Extensions/ApiModelExtensions.cs (hunk 6 above)
Comment (amulware) [reply Toggl.Multivac/Extensions/ApiModelExtensions.cs]: I don't like the shared interface just for this one method, just to then inside check if it's a more specific type. Not to mention that this is only used in a single place as far as I can tell?

File: Toggl.Multivac/Extensions/ApiModelExtensions.cs (hunk 7 above)
Comment (amulware) [reply Toggl.Multivac/Extensions/ApiModelExtensions.cs]: What I don't like is the reflection.

File: Toggl.Multivac/Extensions/ApiModelExtensions.cs (hunk 9 above)
Comment (amulware) [reply Toggl.Multivac/Extensions/ApiModelExtensions.cs]: I think I'm not fully understanding how this is used.

Is the following correct?

```
original = realm.Get()
updatedFromBackend = push(original) // might take a long time
realm.transaction {
  current = realm.Get()
  if original == current // this is where we need this method
    realm.Overwrite(updatedFromBackend)
}
```

File: Toggl.Multivac/Extensions/ApiModelExtensions.cs (hunk 7 above)
Comment (kornelijepetak) [reply Toggl.Multivac/Extensions/ApiModelExtensions.cs]: Yeah... Me neither. Can't see how to do a generic one without reflection here... T4?

Comment (amulware) [reply Toggl.Multivac/Extensions/ApiModelExtensions.cs]: The extension method is only called from a single place, where we generically know the type of the model. Why not make the extension method generic and check the type? Or just use pattern matching to check the type, and then... have generated/hand written comparison methods for the types that don't have at dates? -.-

Completely other proposal: Can we have all IDatabaseSyncable models implement the `At` property?

Comment (simonrozsival) [reply Toggl.Multivac/Extensions/ApiModelExtensions.cs]: You are right that it is now used only from one place - originally it was used in several places.

We could hand write the comparison. Sure.

Putting `At` into all syncable entities - well, no - `IDatabasePreferences`.

File: Toggl.Multivac/Extensions/ApiModelExtensions.cs (hunk 9 above)
Comment (simonrozsival) [reply Toggl.Multivac/Extensions/ApiModelExtensions.cs]: That is correct.

File: Toggl.Foundation.Tests/Toggl.Foundation.Tests.csproj (hunk 15 above)
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Foundation.Tests/Toggl.Foundation.Tests.csproj]: I guess I removed BOM. I won't bother adding it back, sorry.

File: Toggl.Foundation/Sync/ConflictResolution/IConflictResolver.cs [hunk 17]
```diff
@@ -2,7 +2,7 @@
 
 namespace Toggl.Foundation.Sync.ConflictResolution
 {
-    interface IConflictResolver<T>
+    public interface IConflictResolver<T>
```
[the remark this answers is not in the document]
Comment (kornelijepetak) [reply Toggl.Foundation/Sync/ConflictResolution/IConflictResolver.cs]: But why?

File: Toggl.Foundation/Sync/States/Push/DeleteLocalEntityState.cs (hunk 1 above)
Comment (kornelijepetak) [reply Toggl.Foundation/Sync/States/Push/DeleteLocalEntityState.cs]: Yes, you are right. The dangerous world of structs that implement interfaces.

Comment (kornelijepetak) [reply Toggl.Foundation/Sync/States/Push/DeleteLocalEntityState.cs]: Have you forgotten to add a constraint here? Or did I get something wrong?

**Comment by kornelijepetak** [comment]
Should I take it you have decided not to change the name to `CollectionDataSource`?

**Comment by simonrozsival** [comment]
Yes, the renaming I used was like this:
- `IDataSource` => `IBaseDataSource`
- `ISingleObjectDataSource` => `ISingletonDataSource`
- `IManyObjectsDataSource` => `IDataSource`

File: Toggl.Foundation/Sync/ConflictResolution/OverwriteIfLocalDidNotChange.cs [hunk 18]
```diff
@@ -1,23 +1,33 @@
...
-            => localEntity.HasChanged(currentLocal)
+            => hasChanged(currentLocal, localEntity)
                 ? ConflictResolutionMode.Ignore
                 : ConflictResolutionMode.Update;
+
+        private bool hasChanged(TDatabase currentLocal, TDatabase originalLocal)
+            => currentLocal is ILastChangedDatable currentLastChangedDatable &&
+               originalLocal is ILastChangedDatable originalLastChangedDatable
+                ? currentLastChangedDatable.At != originalLastChangedDatable.At
+                : !noPropertyChanged(currentLocal, originalLocal);
```
Comment (amulware) [inline Toggl.Foundation/Sync/ConflictResolution/OverwriteIfLocalDidNotChange.cs]: Double negation, please change method name to `allPropertiesTheSame` or similar.

File: Toggl.Foundation.Tests/Sync/States/Push/PushStateTests.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (amulware) [reply Toggl.Foundation.Tests/Sync/States/Push/PushStateTests.cs]: I mean, if we change, for example, a time entry locally, do we change the `at` date to 'now'? every time, i.e. it changes in the database with every change the user makes?

**Comment by amulware** [comment]
The changes are looking alright to me, but I haven't checked f1d6260edde2bc92087afde0efeb4f3a9c34e1e4 in detail yet.

File: Toggl.Foundation.Tests/Sync/States/Push/PushStateTests.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Foundation.Tests/Sync/States/Push/PushStateTests.cs]: Yes, that is exactly what happens.

File: Toggl.Foundation/Sync/ConflictResolution/IConflictResolver.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Foundation/Sync/ConflictResolution/IConflictResolver.cs]: I made a few changes to keep this internal.

**Comment by simonrozsival** [comment]
About the last changes:
- the data source doesn't expose conflict and rivals resolvers anymore
- I added the extension method back - it is used from two places, so it is useful - it is now based on database models, which actually makes sense
- the conflict resolver can now be internal anymore

File: Toggl.Foundation/DataSources/SingletonDataSource.cs [hunk 19]
```diff
@@ -53,35 +51,24 @@ public override IObservable<TThreadsafe> Create(TThreadsafe entity)
...
+                    return new UpdateResult<TThreadsafe>(u.OriginalId, u.Entity);
 
-                case UpdateResult<TThreadsafe> updated:
-                    currentSubject.OnNext(updated.Entity);
-                    break;
+                case IgnoreResult<TThreadsafe> i:
+                    return new IgnoreResult<TThreadsafe>(i.Id);
             }
+
+            throw new InvalidOperationException($"Cannot handle conflict resolution result {result.GetType().FullName} in the singleton data source.");
```
Comment (amulware) [inline Toggl.Foundation/DataSources/SingletonDataSource.cs]: Creation is not supported? How do we create the very first object?

File: Toggl.Foundation/DataSources/ObservableDataSource.cs [hunk 20]
```diff
@@ -54,18 +53,14 @@ public override IObservable<Unit> Delete(long id)
...
-            => base.UpdateWithConflictResolution(original, entity, conflictResolver, rivalsResolver)
+        public override IObservable<IConflictResolutionResult<TThreadsafe>> OverwriteIfOriginalDidNotChange(
+            TThreadsafe original, TThreadsafe entity)
+            => base.OverwriteIfOriginalDidNotChange(original, entity)
                 .Do(handleBatchUpdateResult);
 
         public override IObservable<IEnumerable<IConflictResolutionResult<TThreadsafe>>> BatchUpdate(IEnumerable<TThreadsafe> entities)
             => base.BatchUpdate(entities)
                 .Do(updatedEntities => updatedEntities
-                    .Where(result => !(result is IgnoreResult<TThreadsafe>))
```
Comment (amulware) [inline Toggl.Foundation/DataSources/ObservableDataSource.cs]: Why is this removed?

Comment (simonrozsival) [reply Toggl.Foundation/DataSources/ObservableDataSource.cs]: It will be ignored - the switch doesn't have a case for it.

File: Toggl.Foundation/DataSources/SingletonDataSource.cs (hunk 19 above)
Comment (simonrozsival) [reply Toggl.Foundation/DataSources/SingletonDataSource.cs]: 🤔

File: Toggl.Foundation/Extensions/DatabaseModelExtensions.cs [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+    public static class DatabaseModelExtensions
+    {
+        public static bool DiffersFrom<TDatabase>(this TDatabase first, TDatabase second)
+            where TDatabase : IDatabaseModel
+            => first is ILastChangedDatable currentLastChangedDatable &&
+               second is ILastChangedDatable originalLastChangedDatable
+                ? currentLastChangedDatable.At != originalLastChangedDatable.At
+                : anyPropertyChanged(first, second);
+
+        private static bool anyPropertyChanged<TDatabase>(TDatabase second, TDatabase originalLocal)
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Foundation/Extensions/DatabaseModelExtensions.cs]: 🙈

File: Toggl.Foundation/Extensions/DatabaseModelExtensions.cs [hunk 22]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Linq;
+using Toggl.Multivac.Models;
+using Toggl.PrimeRadiant.Models;
+
+namespace Toggl.Foundation.Extensions
+{
+    public static class DatabaseModelExtensions
+    {
+        public static bool DiffersFrom<TDatabase>(this TDatabase first, TDatabase second)
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply Toggl.Foundation/Extensions/DatabaseModelExtensions.cs]: I will rename it to `T`.

File: Toggl.Foundation/DataSources/SingletonDataSource.cs (hunk 19 above)
Comment (simonrozsival) [reply Toggl.Foundation/DataSources/SingletonDataSource.cs]: It must support creation, otherwise it would 💥 I will put it back.

**Comment by amulware** [comment]
f1d6260edde2bc92087afde0efeb4f3a9c34e1e4 hurts my brain a bit, but I think it's alright.
